### PR TITLE
fix issue linux orderly shutdown not noticed by xcat. #1939

### PR DESCRIPTION
**Pull request by immarvin**
fix issue https://github.com/xcat2/xcat-core/issues/1939:
1. replace xcatpostinit1 sysvinit service with xcatpostinit1.service systemd service unit as the hook to run PS&PBS and report node status for redhat;
2. report node status on system <reboot/shutdown> command

with this PR, the node status will be changed to "powering-off" on system "reboot" or "shutdown"

### CODE REVIEW

File: xCAT-server/share/xcat/install/scripts/post.xcat [hunk 1]
```diff
@@ -266,50 +268,60 @@ else
...
-    fi
-else
-    if [ "$XCATDEBUGMODE" = "1" ] || [ "$XCATDEBUGMODE" = "2" ]; then
-        msgutil_r "$MASTER_IP" "debug" "/etc/init.d/xcatpostinit1 generated" "/var/log/xcat/xcat.log"
-    fi
-fi
-
 export OSVER=#TABLE:nodetype:THISNODE:os#
+#create the post init service as a hook to run PS and PBS, as well as status updating
+if [ -e /usr/lib/systemd/  ]; then
```
Comment (daniceexi) [inline xCAT-server/share/xcat/install/scripts/post.xcat]: Is it a right way to check whether the system supports the systemd through checking the dir path '/usr/lib/systemd'?
Is it possible the path of the systemd would be changed someday later?

File: xCAT-server/share/xcat/install/scripts/post.xcat [hunk 2]
```diff
@@ -327,12 +339,13 @@ if [[ $OSVER == ubuntu* ]]; then
         msgutil_r "$MASTER_IP" "debug" "update-rc.d -f xcatpostinit1 remove" "/var/log/xcat/xcat.log"
     fi
 else
-    if [ "$RUNBOOTSCRIPTS" != "'yes'" ]; then
+    if [ "$RUNBOOTSCRIPTS" != "'yes'" ] && [ "$NODESTATUS" = "'n'" ]; then
         chkconfig xcatpostinit1 off
+        if [ "$XCATDEBUGMODE" = "1" ] || [ "$XCATDEBUGMODE" = "2" ]; then
```
Comment (daniceexi) [inline xCAT-server/share/xcat/install/scripts/post.xcat]: Does this ``chkconfig xcatpostinit1 off`` also work for systemd?

File: xCAT/postscripts/xcatpostinit1.install [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/bin/sh
```
Comment (daniceexi) [inline xCAT/postscripts/xcatpostinit1.install]: What's the difference between the 'xcatpostinit1.netboot' and 'xcatpostinit1.install'?

File: xCAT/postscripts/xcatpostinit1.service [hunk 4]
```diff
@@ -0,0 +1,12 @@
+[Unit]
```
Comment (daniceexi) [inline xCAT/postscripts/xcatpostinit1.service]: Will this 'xcatpostinit1.service' be added permanently?

**Comment by whowutwut** [comment]
I think this subject can be reduced for the pull request to only say "fix issue linux orderly shutdown not noticed by xcat." and leave the details for the description.

File: xCAT-server/share/xcat/install/scripts/post.xcat (hunk 2 above)
Comment (immarvin) [inline xCAT-server/share/xcat/install/scripts/post.xcat]: yes, the systemd provided partial of backward compatibility for sysvinit. you can see it in:
````
[root@c910f03c05k21 xcat-core]# chkconfig syslog off
Note: Forwarding request to 'systemctl disable syslog.service'.
````

**Comment by immarvin** [comment]
hi @whowutwut , modified

**Comment by immarvin** [comment]
hi @daniceexi , thanks for you online&offline comments, the PR has been updated and verified on rhels7.3 rc3

File: xCAT/postscripts/xcatpostinit1.install [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  ;;
+start)
+  # check for the REBOOT specified in xcatinfo to run post boot scripts on reboot 
+  if [ -f /opt/xcat/xcatinfo ]; then
+    REBOOT=`grep 'REBOOT' /opt/xcat/xcatinfo |cut -d= -f2`
+  fi
+  # if the xcatdsklspost file exists and this is a reboot - run xcatdsklspost with a mode of 6
+  if [ "$REBOOT" = "TRUE" -a -r /opt/xcat/xcatdsklspost -a "$RUNBOOTSCRIPT" = "1"]; then
+      /opt/xcat/xcatdsklspost 6
+  elif [ "$REBOOT" = "TRUE" -a "NODESTATUS" != "n" ]; then
```
Comment (daniceexi) [inline xCAT/postscripts/xcatpostinit1.install]: I think the condition to ``updateflag`` is not correct. We should do the status update whenever "NODESTATUS" != "n"

File: xCAT/postscripts/xcatpostinit1.install [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      /xcatpost/updateflag.awk $XCATSERVER 3002 "installstatus booted"
```
Comment (daniceexi) [inline xCAT/postscripts/xcatpostinit1.install]: Does the ``NODESTATUS`` attribute only support 'n'/'y'? how about 'N'/'Y' and no,yes?

File: xCAT-server/share/xcat/install/scripts/post.xcat [hunk 7]
```diff
@@ -266,50 +268,60 @@ else
...
+#create the post init service as a hook to run PS and PBS, as well as status updating
+if [ -e /usr/lib/systemd/  ]; then
+    cat >/etc/systemd/system/xcatpostinit1.service <<'EOF'
+#INCLUDE:/install/postscripts/xcatpostinit1.service#
+EOF
+    ln -s /etc/systemd/system/xcatpostinit1.service /etc/systemd/system/multi-user.target.wants/xcatpostinit1.service    
 
-if [[ $OSVER == ubuntu* ]]; then
-    ln -s /etc/init.d/xcatpostinit1 /etc/rc2.d/S84xcatpostinit1
+    cat >/opt/xcat/xcatpostinit1 << 'EOF'
```
Comment (daniceexi) [inline xCAT-server/share/xcat/install/scripts/post.xcat]: You did not cover this comment for your change.

File: xCAT/postscripts/xcatpostinit1.service (hunk 4 above)
Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.service]: yes, but whether to run PBS and whether to report node status is controlled by switch variables in site table

File: xCAT/postscripts/xcatpostinit1.install (hunk 3 above)
Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.install]: the description have been added into the 2 scripts.

File: xCAT/postscripts/xcatpostinit1.install (hunk 6 above)
Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.install]: only "n" is mentioned in manpage of site table, 
````
[root@c910f03c05k21 ~]# tabdump -d site|grep nodestatus -A10
 nodestatus:  If set to 'n', the nodelist.status column will not be updated during
              the node deployment, node discovery and power operations. The default is to update.
````

File: xCAT/postscripts/xcatpostinit1.install (hunk 5 above)
Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.install]: how about leave "site.nodestatus" as a switch variable to control whether to spent  one time http communication with MN for the provisioned node?

Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.install]: maybe this switch can be take as a tunable variable in some scaling environment , what do you thing?

File: xCAT/postscripts/xcatpostinit1.install (hunk 3 above)
Comment (immarvin) [inline xCAT/postscripts/xcatpostinit1.install]: code updated

**Comment by daniceexi** [comment]
I have reviewed the change and agree to merge.

**Pull request merged**